### ISSUE TITLE
Submit button not disabled fully during validation

### ISSUE THREAD

**Issue by brianlenz** [formvalidation/support#150]
When you click the submit button, FormValidation will validate the form. The first thing it does is call the validate method, which calls this.disableSubmitButtons(true);. That's great!

Except...then, when each individual field is validated, it re-enables the submit buttons for each form field that hasn't yet been validated :(

In updateStatus():

```
                        if (isValidField === true) {
                            this.disableSubmitButtons(this.isValid() === false);
```

That effectively calls disableSubmitButtons with false for every form field in the form that is being validated. In our case, we have 73 form fields with validation, so there is a slight delay in processing the validate. The submit button should be disabled throughout that process since the submit button was clicked, so I believe this is a bug with submit handling. Shouldn't the submit button stay disabled during the post-submit validate process? I understand the intent of that code is to disable the submit button if the field is in error, but it seems that behavior should be bypassed during form submit? Only once the complete validate call has completed should the submit button status be updated accordingly to disabled/enabled depending on the outcome of the validation.

It's worth noting that we are not using the built-in behavior of disabling the submit button until all form fields are validated. We're using:

```
$('#${formId}').formValidation()
        .on('err.field.fv', function(e, data) {
            data.fv.disableSubmitButtons(false);
        })
        .on('success.field.fv', function(e, data) {
            data.fv.disableSubmitButtons(false);
        });
```

We generally want the submit button to always be enabled, but we want it disabled once the form is actually submitted (which is handled properly in _onSuccess). I've confirmed that our custom err/success handlers are not the issue; it's the native FormValidation.io code that is causing the submit button to be re-enabled.

Note that if this bug gets fixed, we would then need a way in our handlers to detect if the submit process is happening so that we can bypass our own disableSubmitButtons(false) calls in our err/success handlers.

**Comment by MB34** [comment]
I think you should change the event that you are listening to to `status.field.fv` and check for the 'VALIDATING' value. http://formvalidation.io/settings/#event-field

From reading the documentation, listening to `err.field.fv` only disables the button when a validation fails.

**Comment by brianlenz** [comment]
Thanks for the suggestion, but I believe we're already doing what the documentation suggests:

http://formvalidation.io/examples/enabling-submit-button/

In fact, I didn't mention it here, but in addition to success.form.fv and err.form.fv, we are also listening for status.field.fv. The point of our handlers is to keep the submit button enabled all of the time up until the form is submitted, at which point the library should automatically disable it. I'm still convinced this is a bug in the library itself, as the attempt to disable the submit button (which is the first thing that happens in the call to validate) gets overridden by each individual field as it is getting validated. The library needs to start tracking when it is doing a final validation for form submission so that it bypasses the re-enabling of the submit button.

I guess the one thing I could do as a workaround is add my own submit handler to the form to track when the form is being submitted so that I can keep the button disabled that way. But, I really do feel this is a bug in the library that needs to be fixed more generally.

**Comment by MB34** [comment]
I feel that the submit button should be disabled until all fields have passed validation to prevent multiple submission attempts. If your validation takes any amount of time, especially if you have remote validators, then the user could possibly attempt to submit the form again during the processing of the validation.

**Comment by MB34** [comment]
I think the architecture of the plugin is flawed in this respect and the handling of the submit button should be tied to the validation of the entire form vs. each field.

**Comment by brianlenz** [comment]
That's really a separate issue than what I've reported here.

**Comment by brianlenz** [comment]
Phuoc,

It looks like we can't even add a workaround to ensure the submit button is disabled while the form is being validated. The issue is still the code that I referenced in my original post. I tried to work around it by tracking when the form is being submitted with:

```
var isSubmittingForm = false;
function updateSubmitButtons(e, data) {
    data.fv.disableSubmitButtons(isSubmittingForm);
}
$('#${formId}')
        .on('submit', function(event) {
            isSubmittingForm = true;
        })
        .formValidation()
        .on('err.field.fv', updateSubmitButtons)
        .on('status.field.fv', updateSubmitButtons)
        .on('success.field.fv', updateSubmitButtons)
        .on('err.form.fv', function(event) {
            isSubmittingForm = false;
            $(event.target).data('formValidation').disableSubmitButtons(false);
        });
```

That code would work theoretically, but the issue is still that when each individual field is being validated, it forces the submit button to be enabled.

Additionally, the validate process actually blocks the browser from displaying the submit button as disabled right away, too. Since the JavaScript thread is eating up CPU as it validates all of the fields on the form, the browser can't/won't even update the display of the button until after the validation process has been completed. The issue, I believe, is the event handler won't yield to the browser's render engine to update the interface until after the handler has finished executing. The browser is locked up until the complete validate process (and event handler) is finished.

The ultimate hack that I came up with to work around this issue is to twofold.

First, I updated the handlers to skip the default success.fv handler that FormValidation.io adds in place of our own handler that will first disable the submit button, and then yield by executing a setTimeout. Additionally, I added our own custom submit handler for the form to set a flag to indicate that the form is submitting (with a corollary event handler for err.form.fv when validation fails to unset the flag):

```
function enableSubmitButtons(e, data) {
    // when validating form fields, don't disable the submit button.
    data.fv.disableSubmitButtons(false);
}
$('#${formId}')
        .on('submit', function(event) {
            $(event.target).data('isSubmittingForm', true);
        })
        .formValidation()
        .on('err.field.fv', enableSubmitButtons)
        .on('status.field.fv', enableSubmitButtons)
        .on('success.field.fv', enableSubmitButtons)
        .on('err.form.fv', function(event) {
            var $form = $(event.target);
            // when the form is in error, clear the isSubmittingForm flag
            $form.data('isSubmittingForm', false);
            // and re-enable the submit button
            $form.data('formValidation').disableSubmitButtons(false);
        })
        // remove the default submit.fv handler
        .off('submit.fv')
        .on('submit.fv', function(event) {
            // replace the submit.fv handler
            event.preventDefault();
            var formValidation = $(event.target).data('formValidation');
            // disable submit buttons first
            formValidation.disableSubmitButtons(true);
            // initiate the validate via a setTimeout so that the event handler can yield to the browser's render engine to allow the submit button to actually display disabled
            setTimeout(function() {
                formValidation.validate();
            }, 0);
        });
```

Second, I redefined the disableSubmitButtons method. We are using Foundation, so here's the code I put in place:

```
FormValidation.Framework.Foundation.prototype.disableSubmitButtons = function(disabled) {
    // don't allow the submit button to be enabled while the form is in the process of being submitted.
    if(!disabled && this.$form.data('isSubmittingForm')) {
        return;
    }
    return FormValidation.Base.prototype.disableSubmitButtons.apply(this, [disabled]);
};
```

That allows us to prevent the submit button from being enabled during form submission validation.

This would be a much easier problem to solve if the plugin handled it natively. These would certainly be very nice improvements!

Thoughts?
